### DISCUSSION TITLE
Using esModules from XUL

### DISCUSSION THREAD

**Discussion by siefkenj** [protz/thunderbird-conversations#1576] [Q&A] [answered]
I've been thinking about how more of the code can be switched to esModules. `stubs.xhtml` cannot load javascript as an esModule, but it should be possible to use them from iframes. So, a possibility is that `stubs.xhtml` immediately creates a full-sized iframe. This iframe would then message with its parent for whatever it needed, and it should be able to use esmodules.

Questions I don't know the answer to:
  1. What is the right security policy for this iframe? All the modules we want to load are hosted at `chrome://` urls, I think...
  2. What was the plan to fully switch to a web extension? Was this it all along? Is there another way to load web-extension privileged code and stick it in a XUL `browser` component somehow?
  3. How does `remoteContentPolicy` actually work right now? It's unclear who is managing the permissions. I assume that eventually this will all be controlled via content security policies on iframes?

**Comment by Standard8** [comment, marked as answer]
I think we've made a lot of this simpler now with the recent changes. I'm not sure we need to go to the main iframe route at the moment.

> * What was the plan to fully switch to a web extension? Was this it all along? Is there another way to load web-extension privileged code and stick it in a XUL `browser` component somehow?

The plan is roughly to move to a full WebExtension, however, the stub.xhtml was up for debate as to exactly how. It still isn't clear, but the fact iframes have sandbox polices now, might make it more doable than a few years ago.

The basic requirement for the stub page, is that for the message frame itself, we must disable javascript, and only allow remote images when the user says. There's probably a few other requirements that Thunderbird has, that we'd need to make sure that we can do.

I'm not quite sure we're ready to work on that yet - we need to re-arrange the existing architecture to make things simpler. I'll put some notes in the roadmap thread once I get some thoughts together.

> * How does `remoteContentPolicy` actually work right now? It's unclear who is managing the permissions. I assume that eventually this will all be controlled via content security policies on iframes?

The basic idea is that there's a flag included in the message data, and if that flag is set (or not) then the content security policy built into Thunderbird will handle the remote content appropriately. Thunderbird's security policy is based on the url of the item being loaded.

For Conversations, we're hooking into the fact that the email has remote content, and then set the flag and reload if the user wants to view the images.
